panther_manager/src/safety_manager_node.cpp

```diff
@@ -0,0 +1,299 @@
+}
+
+void SafetyManagerNode::CreateShutdownTree()
+{
+  const auto shutdown_hosts_path = this->get_parameter("shutdown_hosts_path").as_string();
+
+  const std::map<std::string, std::any> shutdown_initial_bb = {
+    {"SHUTDOWN_HOSTS_FILE", shutdown_hosts_path.c_str()},
+  };
+
+  shutdown_config_ = bt_utils::CreateBTConfig(shutdown_initial_bb);
+  shutdown_tree_ = factory_.createTree("Shutdown", shutdown_config_.blackboard);
+  shutdown_bt_publisher_ = std::make_unique<BT::Groot2Publisher>(shutdown_tree_, 7777);
+}
+
+void SafetyManagerNode::BatteryCB(const BatteryStateMsg::SharedPtr battery)
+{
+  const auto battery_status = battery->power_supply_status;
+  const auto battery_health = battery->power_supply_health;
+  safety_config_.blackboard->set<unsigned>("battery_status", battery_status);
+  safety_config_.blackboard->set<unsigned>("battery_health", battery_health);
+
+  // don't update battery temperature if unknown status
+  if (
+    battery_status != BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN &&
+    battery_health != BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN) {
+    battery_temp_ma_->Roll(battery->temperature);
+  }
+
+  safety_config_.blackboard->set<double>("bat_temp", battery_temp_ma_->GetAverage());
+}
+
+void SafetyManagerNode::DriverStateCB(const DriverStateMsg::SharedPtr driver_state)
+{
+  front_driver_temp_ma_->Roll(driver_state->front.temperature);
+  rear_driver_temp_ma_->Roll(driver_state->rear.temperature);
+
+  // to simplify conditions pass only higher temp of motor drivers
+  safety_config_.blackboard->set<double>(
+    "driver_temp",
+    std::max({front_driver_temp_ma_->GetAverage(), rear_driver_temp_ma_->GetAverage()}));
+}
+
+void SafetyManagerNode::EStopCB(const BoolMsg::SharedPtr e_stop)
+{
+  safety_config_.blackboard->set<bool>("e_stop_state", e_stop->data);
+}
+
+void SafetyManagerNode::IOStateCB(const IOStateMsg::SharedPtr io_state)
+{
+  if (io_state->power_button) {
+    ShutdownRobot("Power button pressed");
+  }
+
+  safety_config_.blackboard->set<bool>("aux_state", io_state->aux_power);
+  safety_config_.blackboard->set<bool>("fan_state", io_state->fan);
+}
+
+void SafetyManagerNode::SystemStatusCB(const SystemStatusMsg::SharedPtr system_status)
+{
+  cpu_temp_ma_->Roll(system_status->cpu_temp);
+  safety_config_.blackboard->set<double>("cpu_temp", cpu_temp_ma_->GetAverage());
+}
+
+void SafetyManagerNode::SafetyTreeTimerCB()
+{
+  if (!SystemReady()) {
+    return;
+  }
+
+  safety_tree_status_ = safety_tree_.tickOnce();
+
+  if (safety_tree_status_ == BT::NodeStatus::FAILURE) {
+    RCLCPP_WARN(this->get_logger(), "Safety behavior tree returned FAILURE status");
+  }
+
+  std::pair<bool, std::string> signal_shutdown;
+  if (safety_config_.blackboard->get<std::pair<bool, std::string>>(
+        "signal_shutdown", signal_shutdown)) {
+    if (signal_shutdown.first) {
+      ShutdownRobot(signal_shutdown.second);
+    }
+  }
+}
```

Comment:
I'd like to have straightforward answer to this, but well... I'm not sure how will the tree react if it is accessing BB entry and in the same time topic callback want's to modify it. Another thing is if we are fine with possibility for entry change during tree execution. Would have to dig dipper how bt handles this cases

Reply:
I quickly researched this issue, referring to [this article](https://www.behaviortree.dev/docs/guides/ports_vs_blackboard#summary-never-use-the-blackboard-directly), the most likely answer is that BB will suffer thread safety vulnerability.

In general, I still consider the approach of updating the BlackBoard periodically, instead of reading values via BT nodes less reliable and manageable 🤔 We should discuss this issue and possibly reimplement the concept of BlackBoard and Multithreading.